src/Microsoft.ServiceFabric.Services.Remoting/V2/ActivityIdLogicalCallContext.cs

```diff
@@ -0,0 +1,171 @@
+    /// </summary>
+    internal static class ActivityIdLogicalCallContext
+    {
+        /// <summary>
+        /// Creates the activity.
+        /// Note : You still need to start it.
+        /// </summary>
+        /// <param name="activityMessage">The activity message.</param>
+        /// <returns>Activity.</returns>
+        internal static Activity CreateW3CActivity(string activityMessage = "Create new Activity")
+        {
+            Activity.DefaultIdFormat = ActivityIdFormat.W3C;
+            Activity.ForceDefaultIdFormat = true;
+            var activity = new Activity(activityMessage);
+            return activity;
+        }
+
+        internal static void InjectHeaders(IServiceRemotingRequestMessage remotingRequestRequestMessage)
+        {
+            Activity currentActivity = null;
+
+            if (ActivityIdLogicalCallContext.IsPresent())
+            {
+                currentActivity = ActivityIdLogicalCallContext.Get();
+            }
+            else
+            {
+                // If activity ID is not present in AsyncLocal then there's nothing to inject
+                return;
+            }
+
+            if (currentActivity.IdFormat == ActivityIdFormat.W3C)
+            {
+                // If W3C then pass over ID, TraceState and Baggage
+                remotingRequestRequestMessage.GetHeader().ActivityIdParent = currentActivity.Id.ToString();
+                remotingRequestRequestMessage.GetHeader().ActivityIdTraceStateHeader = currentActivity.TraceStateString;
+            }
+            else
+            {
+                remotingRequestRequestMessage.GetHeader().ActivityRequestId = currentActivity.Id.ToString();
+            }
+
+            foreach (var item in currentActivity.Baggage)
+            {
+                remotingRequestRequestMessage.GetHeader().ActivityIdBaggage.Add(item);
+            }
+        }
+
+        internal static void StartActivity(IServiceRemotingRequestMessage requestMessage, ActivitySource activitySource)
+        {
+            string parentId = null;
+            if (requestMessage.GetHeader().ActivityIdParent != null)
+            {
+                // if W3C
+                Activity.DefaultIdFormat = ActivityIdFormat.W3C;
+                Activity.ForceDefaultIdFormat = true;
+                parentId = requestMessage.GetHeader().ActivityIdParent;
+            }
+            else if (requestMessage.GetHeader().ActivityRequestId != null)
+            {
+                // if old scenario
+                parentId = requestMessage.GetHeader().ActivityRequestId;
+            }
+            else
+            {
+                // if activity not present
+                return;
+            }
+
+            // If there are any Listeners then create Activity using ActivitySource
+            var activity = activitySource.StartActivity("StatefulDatabaseIncomingRemoteCall", ActivityKind.Server, CreateActivityContextFromTraceParent(parentId));

```

Comment:
Does ActivitySource.CreateActivity also gives null activity if no listeners present? If yes, you dont need activitySourceListenerPresent  flag

Reply:
The activity object itself is null. But then later once I start adding its attributes, I need to check later below if the ID is null or this flag. Both have same behavior